src/validation.cpp

```diff
@@ -4174,31 +4174,44 @@ void CChainState::LoadExternalBlockFile(FILE* fileIn, FlatFilePos* dbp)
                     continue;
             } catch (const std::exception&) {
                 // no valid block header found; don't complain
+                // (this happens at the end of every blk.dat file)
                 break;
             }
             try {
-                // read block
+                // read block header
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
                 blkdat.SetLimit(nBlockPos + nSize);
-                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                CBlock& block = *pblock;
-                blkdat >> block;
-                nRewind = blkdat.GetPos();
+                CBlockHeader header;
+                blkdat >> header;
 
-                uint256 hash = block.GetHash();
+                // Read the rest of the block (from the end of its header to the end
+                // of the block) to ensure it's in memory while holding cs_main (below).
+                blkdat.Skip(nBlockPos + nSize - blkdat.GetPos());
+
+                const uint256 hash = header.GetHash();
                 {
                     LOCK(cs_main);
-                    // detect out of order blocks, and store them for later
-                    if (hash != m_params.GetConsensus().hashGenesisBlock && !m_blockman.LookupBlockIndex(block.hashPrevBlock)) {
+                    // Store positions of out of order blocks for later.
+                    if (hash != m_params.GetConsensus().hashGenesisBlock && !m_blockman.LookupBlockIndex(header.hashPrevBlock)) {
                         LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
-                                block.hashPrevBlock.ToString());
+                            header.hashPrevBlock.ToString());
                         if (dbp)
-                            mapBlocksUnknownParent.insert(std::make_pair(block.hashPrevBlock, *dbp));
+                            mapBlocksUnknownParent.insert(std::make_pair(header.hashPrevBlock, *dbp));
+
+                        // Skip the rest of this block; position to the marker before the next block.
+                        nRewind = nBlockPos + nSize;
                         continue;
                     }
 
+                    // This block can be processed immediately; rewind to its start then read it.
+                    blkdat.SetPos(nBlockPos);
+                    std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
+                    CBlock& block = *pblock;
+                    blkdat >> block;
```

Comment:
Thanks for the excellent and helpful review!

> It looks like we're deserializing the header twice now

Yes, but since the header is only 80 bytes (less than 0.01% of a 1 MB block), this likely isn't much of a problem. Besides, I don't think it's possible to deserialize a header, and then deserialize only the _rest_ of the block, and end up with a `CBlock`.

> the amount of code executed under cs_main lock increases, since it is now used during the entire block deserialization ...
> you could release/reacquire the lock before/after the block serialization

Yes, good point, I implemented a patch to not hold `cs_main` while deserializing the block (then it's acquired and released 3 times per block read from disk, rather than once). The result was that the reindex time was slower by 15 minutes -- only about 1% but still measurable (although I don't know how much of that difference is due to random variation). I think not much else is going on during reindex; there are other threads in the system but they're just sleeping, so I don't think it's okay to leave it as it is.